README.rst
================
blockscan-python
================

This package was created in the hope to provide a multi-chain API 
wrapper for the blockscan.com endpoints and reduce the overhead of
using multiple single chain wrapper.

Note
----

The following blockchains from `blockscout.com <https://blockscout.com>`_ are also
provided *BUT* it's because blockscout offers the API for `developers transitioning
their applications from Etherscan to BlockScout <https://blockscout.com/etc/mainnet/api-docs>`_. If they deprecate those API in the 
future so will this package.

- Gnosis (100)
- Ethereum Classic (61)
- Ethereum Classic Mordor (63)
- Ethereum Classic Kotti (6)
- POA Core (99)
- POA Sokol (77)
- Artis Sigma1 (246529)
- RSK (30)
- Hoo Smart Chain (70) (This one is technicaly part of blockscan.com but is built on top of blockscout)




Installation
------------

Blockscan-python can be installed using ``pip`` as follows:

.. code-block:: console

   $ pip install blockscan-python

Usage
-----

Refer to the `full documentation <https://blockscan-python.readthedocs.io>`_
for the list of endpoints and the specific blockscan provider for compatibility.
This package doesn't filter the endpoints on a per-chain basis.
This means that some calls will fails if the provider doesn't support it.


Prerequisite
************

- The chain ID you want to connect to.
        Refer to `chainlist <https://chainlist.org/>`_ if unknown
- Specific chain API Token provided by `blockscan <https://blockscan.com>`_
       (i.e: An `etherscan.io <https://etherscan.io>`_ token for Ethereum)

       Exception: Any Blockscout chain takes an empty string "", No token required

Create a connection client
**************************

The connection can be created as *Sync*:

.. code-block:: python

        >>> from blockscan import Blockscan
        
        # Sync connection to etherscan.io
        >>> client = blockscan(1, "MYAPITOKEN", is_async=False)
        >>> client.accounts.get_currency_balance(
                "0x0000000000000000000000000000000000000000"
                )
        '11400022397988649428803'

But will default to *Async* if no parameter is provided:

.. code-block:: python

        >>> from blockscan import BlockScan
        
        #Async connection to bscscan.com
        >>> client = Blockscan(56, "MYAPITOKEN")
        >>> bal = await client.accounts.get_currency_balance(
                        "0x0000000000000000000000000000000000000000"
                        )
        >>> print(bal)
        '1073557893975925234717'

The full list of parameters is available in the `documentation <https://blockscan-python.readthedocs.io/usage.html#create-a-connection>`_



Donate 
------

I made this has a fun side project and it's free for anyone to use.
If you like it and wish to donate here's a few of my crypto wallets. 

.. _tbl-grid:

+----------------------------------------+--------------------------------------+-----------------------------------------+
| Ethereum and L2s (0x29006...)          | Monero (85tBS7YSrM5...)              | Peercoin (PBzj1ZwMDW...)                |
|                                        |                                      |                                         |
+========================================+======================================+=========================================+
| |EthereumQR|                           | |MoneroQR|                           | |PeercoinQR|                            |
+----------------------------------------+--------------------------------------+-----------------------------------------+

.. |EthereumQR| image:: https://raw.githubusercontent.com/SpeakinTelnet/blockscan-python/master/docs/_qrcodes/ethereum.png
  :width: 300
  :alt: EthereumQR

.. |MoneroQR| image:: https://raw.githubusercontent.com/SpeakinTelnet/blockscan-python/master/docs/_qrcodes/monero.png
  :width: 300
  :alt: MoneroQR

.. |PeercoinQR| image:: https://raw.githubusercontent.com/SpeakinTelnet/blockscan-python/master/docs/_qrcodes/peercoin.png
  :width: 300
  :alt: PeerCoinQR

* Free software: MIT license
* Documentation: https://blockscan-python.readthedocs.io.

Credits
-------

Credits to `@pcko1 <https://github.com/pcko1>`_ for making the 
bscscan-python that was used as the base for this package

TO-DOs
------

- Add a reference table of the available Chain ID to the docs


HISTORY.rst
=======
History
=======

0.0.1 (2022-07-02)
------------------

* First release on PyPI.

1.1.0 (2022-07-03)
------------------

* CI tests and deployment added
* Packaging build tested and beta ready

1.1.1 (2022-07-26)
------------------

* Preloaded modules for easier IDE usage


blockscan/blockchains.py
blockchains = {
    # For pytest mocking
    1000000000000: {
        "url": "https://api.test.io/api?",
        "price": "ethprice",
        "supply": "ethsupply",
        "daily_price": "ethdailyprice",
        "daily_market_cap": "ethdailymarketcap",
    },
    #  ETHEREUM
    1: {
        "url": "https://api.etherscan.io/api?",
        "price": "ethprice",
        "supply": "ethsupply",
        "daily_price": "ethdailyprice",
        "daily_market_cap": "ethdailymarketcap",
    },
    # ETHEREUM Ropsten
    3: {
        "url": "https://api-Ropsten.etherscan.io/api?",
        "price": "ethprice",
        "supply": "ethsupply",
        "daily_price": "ethdailyprice",
        "daily_market_cap": "ethdailymarketcap",
    },
    # ETHEREUM Rinkeby
    4: {
        "url": "https://api-Rinkeby.etherscan.io/api?",
        "price": "ethprice",
        "supply": "ethsupply",
        "daily_price": "ethdailyprice",
        "daily_market_cap": "ethdailymarketcap",
    },
    # ETHEREUM Gorli
    5: {
        "url": "https://api-Goerli.etherscan.io/api?",
        "price": "ethprice",
        "supply": "ethsupply",
        "daily_price": "ethdailyprice",
        "daily_market_cap": "ethdailymarketcap",
    },
    # Ethereum Classic Kotti
    6: {
        "url": "https://blockscout.com/etc/kotti/api?",
        "price": "coinprice",
        "supply": "coinsupply",
        "daily_price": "coindailyprice",  # not available
        "daily_market_cap": "coindailymarketcap",  # not available
    },
    # Optimism
    10: {
        "url": "https://api-optimistic.etherscan.io/api?",
        "price": "optimismprice",  # not available
        "supply": "optimismsupply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # CRONOS
    25: {
        "url": "https://api.cronoscan.com/api?",
        "price": "price",  # not available
        "supply": "supply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # RSK
    30: {
        "url": "https://blockscout.com/rsk/mainnet/api?",
        "price": "coinprice",
        "supply": "coinsupply",
        "daily_price": "coindailyprice",  # not available
        "daily_market_cap": "coindailymarketcap",  # not available
    },
    # ETHEREUM Kovan
    42: {
        "url": "https://api-Kovan.etherscan.io/api?",
        "price": "ethprice",
        "supply": "ethsupply",
        "daily_price": "ethdailyprice",
        "daily_market_cap": "ethdailymarketcap",
    },
    # Binance Smart Chain
    56: {
        "url": "https://api.bscscan.com/api?",
        "price": "bnbprice",
        "supply": "bnbsupply",
        "daily_price": "bnbdailyprice",
        "daily_market_cap": "bnbdailymarketcap",  # not available
    },
    # Ethereum Classic
    61: {
        "url": "https://blockscout.com/etc/mainnet/api?",
        "price": "coinprice",
        "supply": "coinsupply",
        "daily_price": "coindailyprice",  # not available
        "daily_market_cap": "coindailymarketcap",  # not available
    },
    # Ethereum Classic Mordor
    63: {
        "url": "https://blockscout.com/etc/mordor/api?",
        "price": "coinprice",
        "supply": "coinsupply",
        "daily_price": "coindailyprice",  # not available
        "daily_market_cap": "coindailymarketcap",  # not available
    },
    # Optimism kovan
    69: {
        "url": "https://api-kovan-optimistic.etherscan.io/api?",
        "price": "optimismprice",  # not available
        "supply": "optimismsupply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # Hoo Smart Chain
    70: {
        "url": "https://hooscan.com/api?",
        "price": "coinprice",
        "supply": "coinsupply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # POA Sokol
    77: {
        "url": "https://blockscout.com/poa/sokol/api?",
        "price": "coinprice",
        "supply": "coinsupply",
        "daily_price": "coindailyprice",  # not available
        "daily_market_cap": "coindailymarketcap",  # not available
    },
    # Binance Smart Chain Testnet
    97: {
        "url": "https://api-testnet.bscscan.com/api?",
        "price": "bnbprice",
        "supply": "bnbsupply",
        "daily_price": "bnbdailyprice",
        "daily_market_cap": "bnbdailymarketcap",  # not available
    },
    # POA Core
    99: {
        "url": "https://blockscout.com/poa/core/api?",
        "price": "coinprice",
        "supply": "coinsupply",
        "daily_price": "coindailyprice",  # not available
        "daily_market_cap": "coindailymarketcap",  # not available
    },
    # Gnosis
    100: {
        "url": "https://blockscout.com/xdai/mainnet/api?",
        "price": "coinprice",
        "supply": "coinsupply",
        "daily_price": "coindailyprice",  # not available
        "daily_market_cap": "coindailymarketcap",  # not available
    },
    # Huobi ECO
    128: {
        "url": "https://api.hecoinfo.com/api?",
        "price": "htprice",
        "supply": "htsupply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # Polygon
    137: {
        "url": "https://api.polygonscan.com/api?",
        "price": "maticprice",
        "supply": "maticsupply",
        "daily_price": "ethdailyprice",
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # BitTorrent
    199: {
        "url": "https://api.bttcscan.com/api?",
        "price": "price",  # not available
        "supply": "supply",
        "daily_price": "ethdailyprice",
        "daily_market_cap": "ethdailymarketcap",
    },
    # Fantom
    250: {
        "url": "https://api.ftmscan.com/api?",
        "price": "ftmprice",
        "supply": "ftmsupply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # Huobi ECO Testnet
    256: {
        "url": "https://api-testnet.hecoinfo.com/api?",
        "price": "htprice",
        "supply": "htsupply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # Cronos Testnet
    338: {
        "url": "https://api-testnet.cronoscan.com/api?",
        "price": "price",  # not available
        "supply": "supply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # CLV Parachain
    1024: {
        "url": "https://api.clvscan.com/api?",
        "price": "price",  # not available
        "supply": "supply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # BitTorrent Testnet
    1028: {
        "url": "https://api-testnet.bttcscan.com/api?",
        "price": "price",  # not available
        "supply": "supply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    #  Moonbeam
    1284: {
        "url": "https://api-moonbeam.moonscan.io/api?",
        "price": "price",  # not available
        "supply": "supply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    #  Moonriver
    1285: {
        "url": "https://api-moonriver.moonscan.io/api?",
        "price": "price",  # not available
        "supply": "supply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    #  Moonbase
    1287: {
        "url": "https://api-moonbase.moonscan.io/api?",
        "price": "price",  # not available
        "supply": "supply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # Fantom Testnet
    4002: {
        "url": "https://api-testnet.ftmscan.com/api?",
        "price": "ftmprice",
        "supply": "ftmsupply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # Arbitrum One
    42161: {
        "url": "https://api.arbiscan.io/api?",
        "price": "price",  # not available
        "supply": "supply",  # not available
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # Celo
    42220: {
        "url": "https://api.celoscan.xyz/api?",
        "price": "price",  # not available
        "supply": "supply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    #  Avalanche Fuji Testnet
    43113: {
        "url": "https://api-testnet.snowtrace.io/api?",
        "price": "price",  # not available
        "supply": "supply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # Avalanche
    43114: {
        "url": "https://api.snowtrace.io/api?",
        "price": "price",  # not available
        "supply": "supply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # Celo Alfajores
    44787: {
        "url": "https://api-alfajores.celoscan.xyz/api?",
        "price": "price",  # not available
        "supply": "supply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # Polygon Mumbai
    80001: {
        "url": "https://api-testnet.polygonscan.com/api?",
        "price": "maticprice",
        "supply": "maticsupply",
        "daily_price": "ethdailyprice",
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # Artis sigma1
    246529: {
        "url": "https://blockscout.com/artis/sigma1/api?",
        "price": "coinprice",
        "supply": "coinsupply",
        "daily_price": "coindailyprice",  # not available
        "daily_market_cap": "coindailymarketcap",  # not available
    },
    # Arbitrum Rinkeby
    421611: {
        "url": "https://api-testnet.arbiscan.io/api?",
        "price": "price",  # not available
        "supply": "supply",  # not available
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # ETHEREUM Sepolia
    11155111: {
        "url": "https://api-Sepolia.etherscan.io/api?",
        "price": "ethprice",
        "supply": "ethsupply",
        "daily_price": "ethdailyprice",
        "daily_market_cap": "ethdailymarketcap",
    },
    # Aurora
    1313161554: {
        "url": "https://api.aurorascan.dev/api?",
        "price": "price",  # not available
        "supply": "supply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
    # Aurora Testnet
    1313161555: {
        "url": "https://api-testnet.aurorascan.dev/api?",
        "price": "price",  # not available
        "supply": "supply",
        "daily_price": "ethdailyprice",  # not available
        "daily_market_cap": "ethdailymarketcap",  # not available
    },
}
